x/archivedb/key_value_reader.go

```diff
@@ -82,13 +28,44 @@ func (reader *dbHeightReader) Has(key []byte) (bool, error) {
 }
 
 // Get retrieves the given key if it's present in the key-value data store.
+//
+// This is a public API, so the dbKey is used to retrieve any value. It is
+// guaranteed to be a O(1), because how the dbKey is structured internally which
+// prevents a longer key to match a shorter key with the same prefix.
+//
+// If the result matches a dbMetaKey it will be consired as a ErrNotFound
+// because dbMetaKey is for internal usage and should be leaked outside of the
+// module
 func (reader *dbHeightReader) Get(key []byte) ([]byte, error) {
-	iterator, err := reader.getIteratorToValue(key)
-	defer iterator.Release()
+	iterator := reader.db.rawDB.NewIteratorWithStart(newKey(key, reader.height).Bytes())
+
+	reader.heightLastFoundKey = 0
+
+	if !iterator.Next() {
+		// There is no available key with the requested prefix
+		return nil, database.ErrNotFound
```

Comment:
should check `iterator.Error()` here as well. (similar comment on the base PR)

Reply:
We should check the value of `iterator.Error` if `Next()` returns false.
This is to distinguish whether there was an error in iteration vs the end of the keyspace.
If .Next() returns true, the error will be nil.